### DISCUSSION TITLE
Cleaning up resources in Tor

### DISCUSSION THREAD

**Discussion by ghost** [shadow/shadow#1966] [Q&A] [answered]
I am working on a project that requires cell-level statistics for each Tor process being run by Shadow. I have accomplished this by modifying Tor to create a file for each channel (circuit) with stdio, writing a line every time a cell is sent or received. However, there appears to be no reliable way to close these files. Tor does not return from main, and most channels remain open when it exits. This results in consistent loss of data (whatever is in the stdio buffers), even causing some files to end up with no content. Is there any recommended way to clean up such resources in Tor when it is used with Shadow?

**Comment by stevenengler** [comment, marked as answer]
When the simulation ends Shadow sends a SIGKILL to all processes, so tor isn't able to flush any stdio buffers to the files. This is something we'd [like to improve](https://github.com/shadow/shadow/issues/1491) in the future. A workaround for the current shadow main branch is to manually send a SIGTERM to each tor process before the simulation ends. For example, if you had the following simulation that ends at 180 seconds:

```yaml
general:
  stop_time: 180s
hosts:
  4uthority:
    network_node_id: 0
    ip_addr: 100.0.0.1
    processes:
    - path: ~/.local/bin/tor
      args: --Address 4uthority --Nickname 4uthority --defaults-torrc torrc-defaults -f torrc
  exit1:
    network_node_id: 0
    processes:
    - path: ~/.local/bin/tor
      args: --Address exit1 --Nickname exit1 --defaults-torrc torrc-defaults -f torrc
      start_time: 60
...
```

You could add a `kill -SIGTERM <pid>` command for each tor process on each host. For example:

```yaml
general:
  stop_time: 180s
hosts:
  4uthority:
    network_node_id: 0
    ip_addr: 100.0.0.1
    processes:
    - path: ~/.local/bin/tor
      args: --Address 4uthority --Nickname 4uthority --defaults-torrc torrc-defaults -f torrc
    - path: /usr/bin/kill
      args: -SIGTERM 1000
      start_time: 179s
  exit1:
    network_node_id: 0
    processes:
    - path: ~/.local/bin/tor
      args: --Address exit1 --Nickname exit1 --defaults-torrc torrc-defaults -f torrc
      start_time: 60
    - path: /usr/bin/kill
      args: -SIGTERM 1000
      start_time: 179s
...
```

This relies on a Shadow implementation detail where Shadow gives processes on each host pids starting at 1000. This may change in the future, but should work for now until we have a better workaround. Processes are assigned pids according to the order in which they are listed in the configuration file.

Below is an example of what pids are assigned to which processes.

```yaml
hosts:
  client1:
    processes:
    - path: /bin/a # given pid 1000
      start_time: 1s
    - path: /bin/b # given pid 1001
      start_time: 3s
    - path: /bin/c # given pid 1002
      start_time: 2s
  client2:
    processes:
    - path: /bin/d # given pid 1000
      start_time: 4s
```

To see how the tor process is cleaned up properly, the tor process stdout shows something like the following without the `kill` commands (the stdout log ends abruptly):

```
Jan 01 00:29:59.000 [notice] Average packaged cell fullness: 29.062%. TLS write overhead: 1%
Jan 01 00:29:59.000 [notice] Circuit handshake stats since last time: 0/0 TAP, 0/0 NTor.
Jan 01 00:29:59.000 [notice] Since startup we initiated 0 and received 0 v1 connections; initiated 0 and received 0 v2 connections; initiated 0 and received 0 v3 connections; initiated 0 and received 0 v4 connections; initiated 5 and received 6 v5 connections.
Jan 01 00:29:59.000 [notice] Heartbeat: DoS mitigation since startup: 0 circuits killed with too many cells, [DoSCircuitCreationEnabled disabled], [DoSConnectionEnabled disabled], [DoSRefuseSingleHopClientRendezvous disabled], 0 INTRODUCE2 rejected.
Jan 01 00:29:59.050 [info] circuit_mark_for_close_(): Circuit 3276807138 (id: 0) marked for close at src/core/or/command.c:657 (orig reason: 512, new reason: 0)
Jan 01 00:29:59.050 [info] circuit_free_(): Circuit 0 (id: 0) has been freed.
```

And with the `kill` command the stdout shows:

```
Jan 01 00:29:59.000 [notice] Average packaged cell fullness: 29.062%. TLS write overhead: 1%
Jan 01 00:29:59.000 [notice] Circuit handshake stats since last time: 0/0 TAP, 0/0 NTor.
Jan 01 00:29:59.000 [notice] Since startup we initiated 0 and received 0 v1 connections; initiated 0 and received 0 v2 connections; initiated 0 and received 0 v3 connections; initiated 0 and received 0 v4 connections; initiated 5 and received 6 v5 connections.
Jan 01 00:29:59.000 [notice] Heartbeat: DoS mitigation since startup: 0 circuits killed with too many cells, [DoSCircuitCreationEnabled disabled], [DoSConnectionEnabled disabled], [DoSRefuseSingleHopClientRendezvous disabled], 0 INTRODUCE2 rejected.
Jan 01 00:29:59.000 [notice] Catching signal TERM, exiting cleanly.
Jan 01 00:29:59.000 [info] or_state_save(): Saved state to "./state"
Jan 01 00:29:59.000 [info] circuit_free_(): Circuit 0 (id: 0) has been freed.
Jan 01 00:29:59.000 [info] circuit_free_(): Circuit 0 (id: 0) has been freed.
[...]
Jan 01 00:29:59.000 [info] circuit_free_(): Circuit 0 (id: 0) has been freed.
Jan 01 00:29:59.000 [info] circuit_free_(): Circuit 0 (id: 0) has been freed.
Jan 01 00:29:59.000 [info] tor_lockfile_unlock(): Unlocking "./lock"
```

I tested this with the tor minimal test case, so I think it should be fine in large simulations as well.

**Reply by sporksmith** [reply]
This is the best solution for now. https://github.com/shadow/shadow/issues/1491 is a tracking issue for longer term solutions (that don't rely on the PID numbering implementation details)

**Reply by stevenengler** [reply]
Fixed the pid assignment, where pids are actually assigned in the order that the processes are listed in the configuration file, not the order in which they start. (See #2056).
